walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,97 @@
 
+import com.google.common.collect.ImmutableList;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
 
+/**
+ * Implements the "find a meeting" feature which returns the possible times when
+ * a meeting could be scheduled given information about the meeting and a list
+ * of all the events already scheduled.
+ */
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendees = request.getAttendees();
+
+    // There are no options for a meeting longer than a day.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // Get the list of time ranges where the attendees have scheduled events.
+    List<TimeRange> conflicts = findTimeConflicts(events, attendees);
+    return getPossibleMeetingTimes(conflicts, request);
+  }
+
+  /**
+   * Extracts a list of time ranges where any of the given people have events
+   * scheduled. Returns the list sorted by start time in ascending order.
+   */
+  private ImmutableList<TimeRange> findTimeConflicts(
+      Collection<Event> events, Collection<String> people) {
+    return events.stream()
+        .filter(event -> hasCommonAttendees(event, people))
+        .map(Event::getWhen)
+        .sorted(TimeRange.ORDER_BY_START)
+        .collect(ImmutableList.toImmutableList());
+  }
+
+  /**
+   * Returns true if any of the people are attending the event.
+   */
+  private boolean hasCommonAttendees(Event event, Collection<String> people) {
+    Collection<String> attendees = event.getAttendees();
+    return !Collections.disjoint(attendees, people);
+  }
+
+  /**
+   * Gets the list of time ranges when all attendees are free to have
+   * the requested meeting given a list of conflicting time ranges
+   * that is sorted by start time.
+   */
+  private ImmutableList<TimeRange> getPossibleMeetingTimes(
+      List<TimeRange> conflicts, MeetingRequest request) {
+    long meetingDuration = request.getDuration();
+    ImmutableList.Builder<TimeRange> possibleTimes = ImmutableList.<TimeRange>builder();
+    int freeStart = TimeRange.START_OF_DAY;
+
+    // Add time ranges that are in the gaps between conflicting events.
+    for (int conflictIndex = 0; conflictIndex < conflicts.size(); conflictIndex++) {
+      TimeRange conflict = conflicts.get(conflictIndex);
+      int conflictStart = conflict.start();
+      int conflictEnd = conflict.end();
+
+      // Get the latest end time out of all the time conflicts that overlap with the current one.
+      while (conflictIndex < conflicts.size() - 1
+          && conflict.overlaps(conflicts.get(conflictIndex + 1))) {
+        conflictIndex++;
+        conflict = conflicts.get(conflictIndex);
```

Comment:
nit: I would probably try to rename "conflict" here to overlappingConflict since you already had "conflict" defined above.